pkg/agent/ipassigner/ip_assigner_linux.go

```diff
@@ -211,99 +387,156 @@ func (a *ipAssigner) UnassignIP(ip string) (bool, error) {
+		return false, err
 	}
+	return true, nil
+}
 
-	if utilnet.IsIPv4(parsedIP) && a.arpResponder != nil {
-		if err := a.arpResponder.RemoveIP(parsedIP); err != nil {
-			return false, fmt.Errorf("failed to remove IP %v from ARP responder: %v", ip, err)
-		}
+func (a *ipAssigner) unassign(ip net.IP, subnetInfo *crdv1b1.SubnetInfo) error {
+	as, _ := a.getAssignee(subnetInfo, false)
+	// The assignee doesn't exist, meaning the IP has been unassigned previously.
+	if as == nil {
+		return nil
 	}
-	if utilnet.IsIPv6(parsedIP) && a.ndpResponder != nil {
-		if err := a.ndpResponder.RemoveIP(parsedIP); err != nil {
-			return false, fmt.Errorf("failed to remove IP %v from NDP responder: %v", ip, err)
+	if err := as.unassign(ip, subnetInfo); err != nil {
+		return err
+	}
+	if as.deletable() {
+		klog.InfoS("Deleting VLAN sub-interface", "interface", as.logicalInterface.Name, "vlan", subnetInfo.VLAN)
+		if err := as.destroy(); err != nil {
+			return err
 		}
+		delete(a.vlanAssignees, subnetInfo.VLAN)
 	}
-
-	a.assignedIPs.Delete(ip)
-	return true, nil
+	delete(a.assignedIPs, ip.String())
+	return nil
 }
 
 // AssignedIPs return the IPs that are assigned to the dummy device.
-func (a *ipAssigner) AssignedIPs() sets.Set[string] {
+func (a *ipAssigner) AssignedIPs() map[string]*crdv1b1.SubnetInfo {
 	a.mutex.RLock()
 	defer a.mutex.RUnlock()
 	// Return a copy.
-	return a.assignedIPs.Union(nil)
+	copy := map[string]*crdv1b1.SubnetInfo{}
+	for k, v := range a.assignedIPs {
+		copy[k] = v
+	}
+	return copy
 }
 
-// InitIPs loads the IPs from the dummy device and replaces the IPs that are assigned to it
+// InitIPs loads the IPs from the dummy/vlan devices and replaces the IPs that are assigned to it
 // with the given ones. This function also adds the given IPs to the ARP/NDP responder if
 // applicable. It can be used to recover the IP assigner to the desired state after Agent restarts.
-func (a *ipAssigner) InitIPs(ips sets.Set[string]) error {
-	a.mutex.Lock()
-	defer a.mutex.Unlock()
-	if a.dummyDevice != nil {
-		assigned, err := a.loadIPAddresses()
-		if err != nil {
-			return fmt.Errorf("error when loading IP addresses from the system: %v", err)
-		}
-		for ip := range ips.Difference(assigned) {
-			addr := util.NewIPNet(net.ParseIP(ip))
-			if err := netlink.AddrAdd(a.dummyDevice, &netlink.Addr{IPNet: addr}); err != nil {
-				if !errors.Is(err, unix.EEXIST) {
-					return fmt.Errorf("failed to add IP %v to interface %s: %v", ip, a.dummyDevice.Attrs().Name, err)
-				}
-			}
-		}
-		for ip := range assigned.Difference(ips) {
-			addr := util.NewIPNet(net.ParseIP(ip))
-			if err := netlink.AddrDel(a.dummyDevice, &netlink.Addr{IPNet: addr}); err != nil {
-				if !errors.Is(err, unix.EADDRNOTAVAIL) {
-					return fmt.Errorf("failed to delete IP %v from interface %s: %v", ip, a.dummyDevice.Attrs().Name, err)
-				}
-			}
-		}
+func (a *ipAssigner) InitIPs(desired map[string]*crdv1b1.SubnetInfo) error {
```

Comment:
The previous version of `InitIPs` would acquire the mutex for the entire function execution, but now we acquire the mutex for individual calls to `AssignIP` / `UnassignIP`. I assume this is ok based on how the function is meant to be used (called once for initialization, and not concurrently with other calls), but I just wanted to double-check with you.

Reply:
Your understanding is correct, added a comment to make it clear.
> It's not thread-safe and should only be called once for initialization before calling other methods.